exp/api/v1beta2/awsmachinepool_webhook.go

```diff
@@ -162,4 +163,15 @@ func (r *AWSMachinePool) Default() {
 		log.Info("DefaultCoolDown is zero, setting 300 seconds as default")
 		r.Spec.DefaultCoolDown.Duration = 300 * time.Second
 	}
+	if r.Spec.MixedInstancesPolicy != nil {
+		if r.Spec.MixedInstancesPolicy.InstancesDistribution == nil {
+			// Set AWS Defaults if we have mixed instances but no distribution policy set to avoid unnecessary template churn
```

Comment:
In the PR description, you mention "unbounded" churn. I think what you mean is that, if `InstancesDistribution` is undefined, then creating the launch configuration template fails.

I see that `InstancesDistribution` is not marked optional in the API types.

I think we only default optional fields.

Reply:
> In the PR description, you mention "unbounded" churn. I think what you mean is that, if InstancesDistribution is undefined, then creating the launch configuration template fails.

The behavior I observed was that when InstanceDistribution was nil, the Launch Template was created successfully, but a diff was always detected. So every reconciliation, a new version was created. The launch template was successfully "updated" each time, there was no error in the creation or update processes. I had a launch template on version 7000+ until I stopped the controller, over the course of a few hours.

Reply:
Thanks for clarifying!

I took a closer look at the issue description, and the code that compares the actual and desired MixedInstancesPolicy: https://github.com/kubernetes-sigs/cluster-api-provider-aws//blob/6ff9319973b929cb4a2ffceca37bbb748e048c44/exp/controllers/awsmachinepool_controller.go#L519.

I don't understand how you came up with the default values. Are they the AWS defaults?

Reply:
> I don't understand how you came up with the default values. Are they the AWS defaults?

Yes, I believe so, I got these by examining the response AWS was supplying to CAPA.

> How will this change affect an AutoScalingGroup created by an earlier CAPA version that did not have these defaults?

My understanding is that it won't affect them, because existing resources won't refire the default webhook logic, it only fires when a new resource is created. We would have to move the defaults into the reconcile loop to set them for existing resources, which I am not sure is necessary.

Reply:
These are the AWS Console defaults for

- 100% on-demand instances
![image](https://user-images.githubusercontent.com/3445370/235809290-4653696f-8b1f-4e5b-b58e-42d4264438a4.png)

- 100% spot instances
![image](https://user-images.githubusercontent.com/3445370/235809723-425c918c-04c4-4338-873f-e1cec61e981a.png)

Reply:
> The webhook is called for both create and update. Please see the kubebuilder metadata at the top of the file, namely:

I see. It definitely took me a bit to fully grasp the kubebuilder metadata jargon.

> The default strategy for 100% on-demand is "lowest price," not the default you chose, "prioritized."

I see, checking out the codebase, CAPA only supports prioritized as a strategy for `OnDemandAllocationStrategy`:
```
var (
	// OnDemandAllocationStrategyPrioritized uses the order of instance type overrides
	// for the LaunchTemplate to define the launch priority of each instance type.
	OnDemandAllocationStrategyPrioritized = OnDemandAllocationStrategy("prioritized")
)
```

> The default strategy for 100% spot is "price capacity optimized," a strategy that is not even represented in our types!

:open_mouth: 

Do we need to expand this to include the AWS default strategies for both of these enums? The intent of my change wasn't to add new features... But that explains my odd choices in this PR. I'm already starting to forget why I made some of these decisions. :sweat_smile: 

> But these defaults will also apply to existing AWSMachinePool objects, and if there's a difference in the configuration to the one in the actual Launch Template, the Launch Template will be replaced. That might surprise users. WDYT?

To me this suggests we shouldn't introduce any new strategy types in this PR, since users can not select them today, so it would be surprising to me if a bugfix introduced a new strategy _and started defaulting to this new strategy._ I'd expect CAPA defaults to match whatever strategies are supported at the present. 

However, this is a very weak opinion, I can see the other side easily as well - AWS always supported these, the real surprise is that CAPA didn't, sticking to AWS defaults makes CAPA more self-documenting and easier to learn, etc.

Which outcome do we prefer? I don't have much stake since Indeed only uses the strategies CAPA has exposed presently.

Reply:
> To me this suggests we shouldn't introduce any new strategy types in this PR, since users can not select them today, so it would be surprising to me if a bugfix introduced a new strategy and started defaulting to this new strategy.

I agree with you 100%.

Reply:
In that case, are we comfortable leaving these as the defaults? I think the risk for "updating" existing resources is small, because of the two if-guards for the defaulting here. If both conditions are true, and defaults are unset, the resource is in a broken state on the current release. So defaulting it would fix the resource.

Reply:
> One option is to set defaults in the controller, not the webhook. That is, if MixedInstancesPolicy is undefined, set the values using the configuration returned by AWS, and update the AWSMachinePool resource.

This approach sounds fine to me, but it will require adding the missing AWS strategies so they can map correctly to CAPA values.

Reply:
> This approach sounds fine to me, but it will require adding the missing AWS strategies so they can map correctly to CAPA values.

Thanks for pointing that out (again). I keep forgetting that.

Let's suppose we want to add these missing strategies.

Is that a breaking change? I _think_ it is, because an AWSMachinePool resource with a new strategy would be incompatible with a CRD that doesn't support the strategy.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility

Reply:
> If both conditions are true, and defaults are unset, the resource is in a broken state on the current release. So defaulting it would fix the resource.

That's true? You mean, in the current release, AWSMachinePool with `mixedInstancesPolicy` set, but `mixedInstancesPolicy.InstancesDistribution` not set is broken?

I have to agree: If such an AWSMachinePool does not result in a valid Launch Template in the current release,, then we do not need to worry about defaults causing an existing Launch Template to be replaced.

Reply:
What about this option:

If the `mixedInstancesPolicy` is not nil, but `mixedInstancesPolicy.InstancesDistribution` is nil, then we do nothing (same as prior to this PR) in the webhook.

Instead, we change the controller so that it skips the comparison on https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4194/files#diff-9c6c4cb74d0076521d841bb705940e29b5b7b6f81a7c7af31f8586616beec425R551.

Reply:
> That's true? You mean, in the current release, AWSMachinePool with mixedInstancesPolicy set, but mixedInstancesPolicy.InstancesDistribution not set is broken?

Sorta. If you have instance refreshes enabled, it's definitely broken, you will be cycling ec2 nodes endlessly. If you have instance refreshes off, you have version churn, but I believe the machine pool will deploy the ec2 nodes correctly because the asg will be pinned to the first version, but that makes it immutable.

We discovered this because we are testing Machine Pools with mixed instances of m6a and m6i's as a secondary type.

> Instead, we change the controller so that it skips the comparison on https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4194/files#diff-9c6c4cb74d0076521d841bb705940e29b5b7b6f81a7c7af31f8586616beec425R551.

Okay, yeah I like that approach. I think that will work and be fully compatible with existing users.